freqtrade/rpc/rpc.py

```diff
@@ -1115,31 +1115,72 @@ def _rpc_delete(self, trade_id: int) -> dict[str, str | int]:
-            if data:
-                custom_data = [data]
+    def _rpc_list_custom_data(
+        self, trade_id: int | None = None, key: str | None = None, limit: int = 100, offset: int = 0
+    ) -> list[dict[str, Any]]:
+        """
+        Fetch custom data for a specific trade, or all open trades if `trade_id` is not provided.
+        Pagination is applied via `limit` and `offset`.
+
+        Returns an array of dictionaries, each containing:
+        - "trade_id": the ID of the trade (int)
+        - "custom_data": a list of custom data dicts, each with the fields:
+                "id", "ft_trade_id", "cd_key", "cd_type", "cd_value", "created_at", "updated_at"
+        """
+        trades: Sequence[Trade]
+        if trade_id is None:
+            # Get all open trades
+            trades = Trade.session.scalars(
+                Trade.get_trades_query([Trade.is_open.is_(True)]).limit(limit).offset(offset)
+            ).all()
         else:
-            custom_data = trade.get_all_custom_data()
-        return [
-            {
-                "id": data_entry.id,
-                "ft_trade_id": data_entry.ft_trade_id,
-                "cd_key": data_entry.cd_key,
-                "cd_type": data_entry.cd_type,
-                "cd_value": data_entry.cd_value,
-                "created_at": data_entry.created_at,
-                "updated_at": data_entry.updated_at,
-            }
-            for data_entry in custom_data
-        ]
+            trades = Trade.get_trades(trade_filter=[Trade.id == trade_id]).all()
+
+        if not trades:
+            raise RPCException(
+                f"No trade found for trade_id: {trade_id}" if trade_id else "No open trades found."
+            )
+
+        results = []
+        for trade in trades:
+            # Depending on whether a specific key is provided, retrieve custom data accordingly.
+            if key:
+                data = trade.get_custom_data(key=key, retrieval_mode="object")
+                # If data exists, wrap it in a list so the output remains consistent.
+                custom_data = [data] if data else []
+            else:
+                custom_data = trade.get_all_custom_data()
+
+            # Format each custom data entry.
+            formatted_custom_data = [
+                {
+                    "id": data_entry.id,
+                    "ft_trade_id": data_entry.ft_trade_id,
+                    "cd_key": data_entry.cd_key,
+                    "cd_type": data_entry.cd_type,
+                    "cd_value": data_entry.cd_value,
+                    "created_at": data_entry.created_at,
+                    "updated_at": data_entry.updated_at,
+                }
+                for data_entry in custom_data
+            ]
+
+            # Append result for the trade if any custom data was found.
+            if formatted_custom_data:
+                results.append({"trade_id": trade.id, "custom_data": formatted_custom_data})
```

Comment:
This section seems odd
why would you first try to create an object and then add it?
Can't the "if" be moved to the top of the `formatted_custom_data` section, skipping that section if nothing is found?